# Use the ID3 algorithm for the following question. <a class="tocSkip">

# Build Decision Tree

**Build a decision tree from the given tennis dataset. You should build a tree to predict PlayTennis, based on the
other attributes (but, do not use the Day attribute in your tree.). Show all of your work, calculations, and
decisions as you build the tree.
What is the classification accuracy?**

<img src="https://i.gyazo.com/d56d8220c80e3416177b9945203259e8.png">

We will fill the dataset manually since its small and its easter break.

In [43]:
import pandas as pd
import numpy as np

df = pd.DataFrame(columns=["Day","Outlook", "Temperature","Humidity","Wind","PlayTennis"],
                   data=[['D1','Sunny','Hot','High','Weak','No'],
                         ['D2','Sunny','Hot','High','Strong','No'],
                         ['D3','Overcast','Hot','High','Weak','Yes'],
                         ['D4','Rain','Mild','High','Weak','Yes'],
                         ['D5','Rain','Cool','Normal','Weak','Yes'],
                         ['D6','Rain','Cool','Normal','Strong','No'],
                         ['D7','Overcast','Cool','Normal','Strong','Yes'],
                         ['D8','Sunny','Mild','High','Weak','No'],
                         ['D9','Sunny','Cool','Normal','Weak','Yes'],
                         ['D10','Rain','Mild','Normal','Weak','Yes'],
                         ['D11','Sunny','Mild','Normal','Strong','Yes'],
                         ['D12','Overcast','Mild','High','Strong','Yes'],
                         ['D13','Overcast','Hot','Normal','Weak','Yes'],
                         ['D14','Rain','Mild','High','Strong','No']])
df

Day   Outlook Temperature Humidity    Wind PlayTennis
0    D1     Sunny         Hot     High    Weak         No
1    D2     Sunny         Hot     High  Strong         No
2    D3  Overcast         Hot     High    Weak        Yes
3    D4      Rain        Mild     High    Weak        Yes
4    D5      Rain        Cool   Normal    Weak        Yes
5    D6      Rain        Cool   Normal  Strong         No
6    D7  Overcast        Cool   Normal  Strong        Yes
7    D8     Sunny        Mild     High    Weak         No
8    D9     Sunny        Cool   Normal    Weak        Yes
9   D10      Rain        Mild   Normal    Weak        Yes
10  D11     Sunny        Mild   Normal  Strong        Yes
11  D12  Overcast        Mild     High  Strong        Yes
12  D13  Overcast         Hot   Normal    Weak        Yes
13  D14      Rain        Mild     High  Strong         No

In [161]:
#number of positive and negative examples in entire dataset
p=np.sum(df['PlayTennis']=='Yes')
n=np.sum(df['PlayTennis']=='No')
total= n + p
print(p)
print(n)
print(total)

9
5
14


Entropy of entire dataset S

In [162]:
EntropyS= -p/(total)*np.log2(p/(total)) - n/(total)*np.log2(n/(total))
EntropyS

0.9402859586706311

## Choose root node

At first iteration, we need to know which is best attribute to be chosen as top root in our decision tree. To do that, ID3 will find the best attribute which is has **maximum information gain**.

For each Attribute (we will start with **Outlook**)   
-Calculate Entropy for all possible values (here we have 'Sunny','Rainy','Overcast')

### Outlook

In [163]:
OutlookS=df[df['Outlook']=='Sunny']
OutlookR=df[df['Outlook']=='Rain']
OutlookO=df[df['Outlook']=='Overcast']

display(OutlookS.loc[:,['Outlook','PlayTennis']])
display(OutlookR.loc[:,['Outlook','PlayTennis']])
display(OutlookO.loc[:,['Outlook','PlayTennis']])

Outlook PlayTennis
0    Sunny         No
1    Sunny         No
7    Sunny         No
8    Sunny        Yes
10   Sunny        Yes

Outlook PlayTennis
3     Rain        Yes
4     Rain        Yes
5     Rain         No
9     Rain        Yes
13    Rain         No

Outlook PlayTennis
2   Overcast        Yes
6   Overcast        Yes
11  Overcast        Yes
12  Overcast        Yes

In [164]:
#TO AVOID INFINITE LOG WE WILL ADD A VERY SMALL NUMBER TO IT 
eps=0.0000000001

In [165]:
#Entropy for sunny
pS=np.sum(OutlookS['PlayTennis']=='Yes')
nS=np.sum(OutlookS['PlayTennis']=='No')

ESunny=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
ESunny

0.9709505941661296

In [166]:
#Entropy for rain
pR=np.sum(OutlookR['PlayTennis']=='Yes')
nR=np.sum(OutlookR['PlayTennis']=='No')

ERain=-pR/(pR+nR)*np.log2(pR/(pR+nR) + eps) - nR/(pR+nR)*np.log2(nR/(pR+nR) + eps)
ERain

0.9709505941661296

In [167]:
#Entropy for overcast
pO=np.sum(OutlookO['PlayTennis']=='Yes')
nO=np.sum(OutlookO['PlayTennis']=='No')

EOvercast=-pO/(pO+nO)*np.log2(pO/(pO+nO) + eps) - nO/(pO+nO)*np.log2(nO/(pO+nO) + eps)
EOvercast

-1.4426951601859516e-10

In [168]:
# Calculate average information entropy

IOutlook = (pS+nS)/total*ESunny + (pR+nR)/total*ERain + (pO+nO)/total*EOvercast
IOutlook

0.6935361386488726

In [116]:
# Calculate information gain
IGOutlook=EntropyS-IOutlook
IGOutlook

0.2467498200217585

### Temperature

In [117]:
TempH=df[df['Temperature']=='Hot']
TempM=df[df['Temperature']=='Mild']
TempC=df[df['Temperature']=='Cool']

display(TempH.loc[:,['Temperature','PlayTennis']])
display(TempM.loc[:,['Temperature','PlayTennis']])
display(TempC.loc[:,['Temperature','PlayTennis']])

Temperature PlayTennis
0          Hot         No
1          Hot         No
2          Hot        Yes
12         Hot        Yes

Temperature PlayTennis
3         Mild        Yes
7         Mild         No
9         Mild        Yes
10        Mild        Yes
11        Mild        Yes
13        Mild         No

Temperature PlayTennis
4        Cool        Yes
5        Cool         No
6        Cool        Yes
8        Cool        Yes

In [118]:
#Entropy for hot
pH=np.sum(TempH['PlayTennis']=='Yes')
nH=np.sum(TempH['PlayTennis']=='No')

EHot=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHot

0.9999999997114609

In [119]:
#Entropy for mild
pM=np.sum(TempM['PlayTennis']=='Yes')
nM=np.sum(TempM['PlayTennis']=='No')

EMild=-pM/(pM+nM)*np.log2(pM/(pM+nM) + eps) - nM/(pM+nM)*np.log2(nM/(pM+nM) + eps)
EMild

0.9182958337659504

In [120]:
#Entropy for cool
pC=np.sum(TempC['PlayTennis']=='Yes')
nC=np.sum(TempC['PlayTennis']=='No')

ECool=-pC/(pC+nC)*np.log2(pC/(pC+nC) + eps) - nC/(pC+nC)*np.log2(nC/(pC+nC) + eps)
ECool

0.8112781241705938

In [121]:
# Calculate average information entropy

ITemp= (pH+nH)/total*EHot + (pM+nM)/total*EMild + (pC+nC)/total*ECool
ITemp

0.9110633927231372

In [122]:
# Calculate information gain
IGTemp=EntropyS-ITemp
IGTemp

0.02922256594749395

### Humidity

In [123]:
HumidH=df[df['Humidity']=='High']
HumidN=df[df['Humidity']=='Normal']

display(HumidH.loc[:,['Humidity','PlayTennis']])
display(HumidN.loc[:,['Humidity','PlayTennis']])

Humidity PlayTennis
0      High         No
1      High         No
2      High        Yes
3      High        Yes
7      High         No
11     High        Yes
13     High         No

Humidity PlayTennis
4    Normal        Yes
5    Normal         No
6    Normal        Yes
8    Normal        Yes
9    Normal        Yes
10   Normal        Yes
12   Normal        Yes

In [127]:
#Entropy for high
pH=np.sum(HumidH['PlayTennis']=='Yes')
nH=np.sum(HumidH['PlayTennis']=='No')

EHigh=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHigh

0.9852281357457124

In [128]:
#Entropy for normal
pN=np.sum(HumidN['PlayTennis']=='Yes')
nN=np.sum(HumidN['PlayTennis']=='No')

ENormal=-pN/(pN+nN)*np.log2(pN/(pN+nN) + eps) - nN/(pN+nN)*np.log2(nN/(pN+nN) + eps)
ENormal

0.5916727782937884

In [129]:
# Calculate average information entropy

IHumid= (pH+nH)/total*EHigh + (pN+nN)/total*ENormal 
IHumid

0.7884504570197504

In [130]:
# Calculate information gain
IGHumid=EntropyS-IHumid
IGHumid

0.15183550165088078

### Wind

In [133]:
WindS=df[df['Wind']=='Strong']
WindW=df[df['Wind']=='Weak']

display(WindS.loc[:,['Wind','PlayTennis']])
display(WindW.loc[:,['Wind','PlayTennis']])

Wind PlayTennis
1   Strong         No
5   Strong         No
6   Strong        Yes
10  Strong        Yes
11  Strong        Yes
13  Strong         No

Wind PlayTennis
0   Weak         No
2   Weak        Yes
3   Weak        Yes
4   Weak        Yes
7   Weak         No
8   Weak        Yes
9   Weak        Yes
12  Weak        Yes

In [134]:
#Entropy for strong
pS=np.sum(WindS['PlayTennis']=='Yes')
nS=np.sum(WindS['PlayTennis']=='No')

EStrong=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
EStrong

0.9999999997114609

In [135]:
#Entropy for weak
pW=np.sum(WindW['PlayTennis']=='Yes')
nW=np.sum(WindW['PlayTennis']=='No')

EWeak=-pW/(pW+nW)*np.log2(pW/(pW+nW) + eps) - nW/(pW+nW)*np.log2(nW/(pW+nW) + eps)
EWeak

0.8112781241705938

In [136]:
# Calculate average information entropy

IWind= (pW+nW)/total*EWeak + (pS+nS)/total*EStrong 
IWind

0.8921589279738225

In [137]:
# Calculate information gain
IGWind=EntropyS-IWind
IGWind

0.04812703069680868

As we can see the attribute that gives us the maximum information gain is **Outlook** with a gain of 0.247  

Therefore, it will be our root node

In [138]:
display(OutlookS.loc[:,['Outlook','PlayTennis']])
display(OutlookR.loc[:,['Outlook','PlayTennis']])
display(OutlookO.loc[:,['Outlook','PlayTennis']])

Outlook PlayTennis
0    Sunny         No
1    Sunny         No
7    Sunny         No
8    Sunny        Yes
10   Sunny        Yes

Outlook PlayTennis
3     Rain        Yes
4     Rain        Yes
5     Rain         No
9     Rain        Yes
13    Rain         No

Outlook PlayTennis
2   Overcast        Yes
6   Overcast        Yes
11  Overcast        Yes
12  Overcast        Yes

Since Outlook=Overcast has only positive examples, we don't need to split the tree anymore at this node.Therefore, it becomes a **leaf node**.  

For the other two values we must keep splitting using the same procedure.

<img src="https://i.gyazo.com/e36a779b306204d5459dd46f28c0fe04.png">

## Outlook = Sunny sub-tree

In [169]:
#Entropy of sunny
ESunny

0.9709505941661296

In [170]:
OutlookS

Day Outlook Temperature Humidity    Wind PlayTennis
0    D1   Sunny         Hot     High    Weak         No
1    D2   Sunny         Hot     High  Strong         No
7    D8   Sunny        Mild     High    Weak         No
8    D9   Sunny        Cool   Normal    Weak        Yes
10  D11   Sunny        Mild   Normal  Strong        Yes

### Humidity

In [171]:
HumidH=OutlookS[OutlookS['Humidity']=='High']
HumidN=OutlookS[OutlookS['Humidity']=='Normal']

display(HumidH.loc[:,['Humidity','PlayTennis']])
display(HumidN.loc[:,['Humidity','PlayTennis']])

Humidity PlayTennis
0     High         No
1     High         No
7     High         No

Humidity PlayTennis
8    Normal        Yes
10   Normal        Yes

In [172]:
#Entropy for high
pH=np.sum(HumidH['PlayTennis']=='Yes')
nH=np.sum(HumidH['PlayTennis']=='No')

EHigh=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHigh

-1.4426951601859516e-10

In [173]:
#Entropy for normal
pN=np.sum(HumidN['PlayTennis']=='Yes')
nN=np.sum(HumidN['PlayTennis']=='No')

ENormal=-pN/(pN+nN)*np.log2(pN/(pN+nN) + eps) - nN/(pN+nN)*np.log2(nN/(pN+nN) + eps)
ENormal

-1.4426951601859516e-10

In [174]:
total=pS+nS

# Calculate average information entropy

IHumid= (pH+nH)/total*EHigh + (pN+nN)/total*ENormal 
IHumid

-1.4426951601859516e-10

In [175]:
# Calculate information gain
IGHumid=ESunny-IHumid
IGHumid

0.9709505943103991

### Wind

In [176]:
WindS=OutlookS[OutlookS['Wind']=='Strong']
WindW=OutlookS[OutlookS['Wind']=='Weak']

display(WindS.loc[:,['Wind','PlayTennis']])
display(WindW.loc[:,['Wind','PlayTennis']])

Wind PlayTennis
1   Strong         No
10  Strong        Yes

Wind PlayTennis
0  Weak         No
7  Weak         No
8  Weak        Yes

In [177]:
#Entropy for strong
pS=np.sum(WindS['PlayTennis']=='Yes')
nS=np.sum(WindS['PlayTennis']=='No')

EStrong=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
EStrong

0.9999999997114609

In [178]:
#Entropy for weak
pW=np.sum(WindW['PlayTennis']=='Yes')
nW=np.sum(WindW['PlayTennis']=='No')

EWeak=-pW/(pW+nW)*np.log2(pW/(pW+nW) + eps) - nW/(pW+nW)*np.log2(nW/(pW+nW) + eps)
EWeak

0.9182958337659504

In [179]:
# Calculate average information entropy

IWind= (pW+nW)/total*EWeak + (pS+nS)/total*EStrong 
IWind

0.9509775001441545

In [181]:
# Calculate information gain
IGWind=ESunny-IWind
IGWind

0.019973094021975113

### Temperature

In [182]:
TempH=OutlookS[OutlookS['Temperature']=='Hot']
TempM=OutlookS[OutlookS['Temperature']=='Mild']
TempC=OutlookS[OutlookS['Temperature']=='Cool']

display(TempH.loc[:,['Temperature','PlayTennis']])
display(TempM.loc[:,['Temperature','PlayTennis']])
display(TempC.loc[:,['Temperature','PlayTennis']])

Temperature PlayTennis
0         Hot         No
1         Hot         No

Temperature PlayTennis
7         Mild         No
10        Mild        Yes

Temperature PlayTennis
8        Cool        Yes

In [183]:
#Entropy for hot
pH=np.sum(TempH['PlayTennis']=='Yes')
nH=np.sum(TempH['PlayTennis']=='No')

EHot=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHot

-1.4426951601859516e-10

In [184]:
#Entropy for mild
pM=np.sum(TempM['PlayTennis']=='Yes')
nM=np.sum(TempM['PlayTennis']=='No')

EMild=-pM/(pM+nM)*np.log2(pM/(pM+nM) + eps) - nM/(pM+nM)*np.log2(nM/(pM+nM) + eps)
EMild

0.9999999997114609

In [187]:
#Entropy for cool
pC=np.sum(TempC['PlayTennis']=='Yes')
nC=np.sum(TempC['PlayTennis']=='No')

ECool=-pC/(pC+nC)*np.log2(pC/(pC+nC) + eps) - nC/(pC+nC)*np.log2(nC/(pC+nC) + eps)
ECool

-1.4426951601859516e-10

In [188]:
# Calculate average information entropy

ITemp= (pH+nH)/total*EHot + (pM+nM)/total*EMild + (pC+nC)/total*ECool
ITemp

0.3999999997980227

In [189]:
# Calculate information gain
IGTemp=ESunny-ITemp
IGTemp

0.5709505943681069

Obviously, **Humidity** gives us the maximum gain so we split at this attribute.  

<img src="https://i.gyazo.com/4d4ce12cfc13618b7952fae9356d8872.png">

## Outlook  = Rain sub-tree

In [190]:
#Entropy of rain
ERain

0.9709505941661296

In [191]:
OutlookR

Day Outlook Temperature Humidity    Wind PlayTennis
3    D4    Rain        Mild     High    Weak        Yes
4    D5    Rain        Cool   Normal    Weak        Yes
5    D6    Rain        Cool   Normal  Strong         No
9   D10    Rain        Mild   Normal    Weak        Yes
13  D14    Rain        Mild     High  Strong         No

In [193]:
total=pR+nR
total

5

### Humidity

In [194]:
HumidH=OutlookR[OutlookR['Humidity']=='High']
HumidN=OutlookR[OutlookR['Humidity']=='Normal']

display(HumidH.loc[:,['Humidity','PlayTennis']])
display(HumidN.loc[:,['Humidity','PlayTennis']])

Humidity PlayTennis
3      High        Yes
13     High         No

Humidity PlayTennis
4   Normal        Yes
5   Normal         No
9   Normal        Yes

In [195]:
#Entropy for high
pH=np.sum(HumidH['PlayTennis']=='Yes')
nH=np.sum(HumidH['PlayTennis']=='No')

EHigh=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHigh

0.9999999997114609

In [196]:
#Entropy for normal
pN=np.sum(HumidN['PlayTennis']=='Yes')
nN=np.sum(HumidN['PlayTennis']=='No')

ENormal=-pN/(pN+nN)*np.log2(pN/(pN+nN) + eps) - nN/(pN+nN)*np.log2(nN/(pN+nN) + eps)
ENormal

0.9182958337659504

In [197]:
# Calculate average information entropy

IHumid= (pH+nH)/total*EHigh + (pN+nN)/total*ENormal 
IHumid

0.9509775001441545

In [199]:
# Calculate information gain
IGHumid=ERain-IHumid
IGHumid

0.019973094021975113

### Wind

In [200]:
WindS=OutlookR[OutlookR['Wind']=='Strong']
WindW=OutlookR[OutlookR['Wind']=='Weak']

display(WindS.loc[:,['Wind','PlayTennis']])
display(WindW.loc[:,['Wind','PlayTennis']])

Wind PlayTennis
5   Strong         No
13  Strong         No

Wind PlayTennis
3  Weak        Yes
4  Weak        Yes
9  Weak        Yes

In [201]:
#Entropy for strong
pS=np.sum(WindS['PlayTennis']=='Yes')
nS=np.sum(WindS['PlayTennis']=='No')

EStrong=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
EStrong

-1.4426951601859516e-10

In [202]:
#Entropy for weak
pW=np.sum(WindW['PlayTennis']=='Yes')
nW=np.sum(WindW['PlayTennis']=='No')

EWeak=-pW/(pW+nW)*np.log2(pW/(pW+nW) + eps) - nW/(pW+nW)*np.log2(nW/(pW+nW) + eps)
EWeak

-1.4426951601859516e-10

In [203]:
# Calculate average information entropy

IWind= (pW+nW)/total*EWeak + (pS+nS)/total*EStrong 
IWind

-1.4426951601859516e-10

In [204]:
# Calculate information gain
IGWind=ERain-IWind
IGWind

0.9709505943103991

### Temperature

In [205]:
TempH=OutlookR[OutlookR['Temperature']=='Hot']
TempM=OutlookR[OutlookR['Temperature']=='Mild']
TempC=OutlookR[OutlookR['Temperature']=='Cool']

display(TempH.loc[:,['Temperature','PlayTennis']])
display(TempM.loc[:,['Temperature','PlayTennis']])
display(TempC.loc[:,['Temperature','PlayTennis']])

Empty DataFrame
Columns: [Temperature, PlayTennis]
Index: []

Temperature PlayTennis
3         Mild        Yes
9         Mild        Yes
13        Mild         No

Temperature PlayTennis
4        Cool        Yes
5        Cool         No

In [206]:
#Entropy for hot
pH=np.sum(TempH['PlayTennis']=='Yes')
nH=np.sum(TempH['PlayTennis']=='No')

EHot=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHot

C:\Users\Paul\.conda\envs\ml\lib\site-packages\ipykernel_launcher.py:5: RuntimeWarning: invalid value encountered in longlong_scalars
  """


nan

In [207]:
#Entropy for mild
pM=np.sum(TempM['PlayTennis']=='Yes')
nM=np.sum(TempM['PlayTennis']=='No')

EMild=-pM/(pM+nM)*np.log2(pM/(pM+nM) + eps) - nM/(pM+nM)*np.log2(nM/(pM+nM) + eps)
EMild

0.9182958337659504

In [208]:
#Entropy for cool
pC=np.sum(TempC['PlayTennis']=='Yes')
nC=np.sum(TempC['PlayTennis']=='No')

ECool=-pC/(pC+nC)*np.log2(pC/(pC+nC) + eps) - nC/(pC+nC)*np.log2(nC/(pC+nC) + eps)
ECool

0.9999999997114609

In [209]:
# Calculate average information entropy

ITemp=  (pM+nM)/total*EMild + (pC+nC)/total*ECool
ITemp

0.9509775001441545

In [210]:
# Calculate information gain
IGTemp=ERain-ITemp
IGTemp

0.019973094021975113

The attribute with the maximum gain is **Wind** so the final decision tree will look like this.  

<img src="https://i.gyazo.com/bfb3d95e9e18f8f0d6ecd8a69ed84572.png">

For the given data set the classification accuracy is 100% since all leaf nodes are pure.

# Include Temperature?

**Is it possible to produce some set of correct training examples that will get the algorihtm to include the
attribute Temperature in the learned tree, even though the true target concept is independent of
Temperature? if no, explain. If yes, give such a set.**

For attribute Temperature to be selected as a split attribute, it must have the highest information gain (or lowest entropy ).  
We can select a subset of our training examples that actually achieves 0 entropy for temperature,making it the only node in the tree.  

Such a subset is D1,D2,D4,D10,D11,D12

In [211]:
df = pd.DataFrame(columns=["Day","Outlook", "Temperature","Humidity","Wind","PlayTennis"],
                   data=[['D1','Sunny','Hot','High','Weak','No'],
                         ['D2','Sunny','Hot','High','Strong','No'],
                         ['D4','Rain','Mild','High','Weak','Yes'],
                         ['D10','Rain','Mild','Normal','Weak','Yes'],
                         ['D11','Sunny','Mild','Normal','Strong','Yes'],
                         ['D12','Overcast','Mild','High','Strong','Yes']])
df

Day   Outlook Temperature Humidity    Wind PlayTennis
0   D1     Sunny         Hot     High    Weak         No
1   D2     Sunny         Hot     High  Strong         No
2   D4      Rain        Mild     High    Weak        Yes
3  D10      Rain        Mild   Normal    Weak        Yes
4  D11     Sunny        Mild   Normal  Strong        Yes
5  D12  Overcast        Mild     High  Strong        Yes

In [212]:
#number of positive and negative examples in entire dataset
p=np.sum(df['PlayTennis']=='Yes')
n=np.sum(df['PlayTennis']=='No')
total= n + p
print(p)
print(n)
print(total)

4
2
6


In [213]:
EntropyS= -p/(total)*np.log2(p/(total)) - n/(total)*np.log2(n/(total))
EntropyS

0.9182958340544896

In [214]:
TempH=df[df['Temperature']=='Hot']
TempM=df[df['Temperature']=='Mild']
TempC=df[df['Temperature']=='Cool']

display(TempH.loc[:,['Temperature','PlayTennis']])
display(TempM.loc[:,['Temperature','PlayTennis']])
display(TempC.loc[:,['Temperature','PlayTennis']])

Temperature PlayTennis
0         Hot         No
1         Hot         No

Temperature PlayTennis
2        Mild        Yes
3        Mild        Yes
4        Mild        Yes
5        Mild        Yes

Empty DataFrame
Columns: [Temperature, PlayTennis]
Index: []

In [215]:
#Entropy for hot
pH=np.sum(TempH['PlayTennis']=='Yes')
nH=np.sum(TempH['PlayTennis']=='No')

EHot=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHot

-1.4426951601859516e-10

In [216]:
#Entropy for mild
pM=np.sum(TempM['PlayTennis']=='Yes')
nM=np.sum(TempM['PlayTennis']=='No')

EMild=-pM/(pM+nM)*np.log2(pM/(pM+nM) + eps) - nM/(pM+nM)*np.log2(nM/(pM+nM) + eps)
EMild

-1.4426951601859516e-10

In [218]:
# Calculate average information entropy

ITemp= (pH+nH)/total*EHot + (pM+nM)/total*EMild 
ITemp

-1.4426951601859514e-10

All other attributes have greater total entropy,making Temperature our root node and since its children nodes are pure ,they become leaf nodes. 

Our tree consists of only one node (Temperature) and consequently ,the decision to play tennis is made is the Temperature is Mild. 

# Train-Test split

**Now, build a tree using only examples D1–D7. What is the classification accuracy for the training set? what is
the accuracy for the test set (examples D8–D14)? explain why you think these are the results.**

In [219]:
df = pd.DataFrame(columns=["Day","Outlook", "Temperature","Humidity","Wind","PlayTennis"],
                   data=[['D1','Sunny','Hot','High','Weak','No'],
                         ['D2','Sunny','Hot','High','Strong','No'],
                         ['D3','Overcast','Hot','High','Weak','Yes'],
                         ['D4','Rain','Mild','High','Weak','Yes'],
                         ['D5','Rain','Cool','Normal','Weak','Yes'],
                         ['D6','Rain','Cool','Normal','Strong','No'],
                         ['D7','Overcast','Cool','Normal','Strong','Yes']])
df

Day   Outlook Temperature Humidity    Wind PlayTennis
0  D1     Sunny         Hot     High    Weak         No
1  D2     Sunny         Hot     High  Strong         No
2  D3  Overcast         Hot     High    Weak        Yes
3  D4      Rain        Mild     High    Weak        Yes
4  D5      Rain        Cool   Normal    Weak        Yes
5  D6      Rain        Cool   Normal  Strong         No
6  D7  Overcast        Cool   Normal  Strong        Yes

In [220]:
#number of positive and negative examples in entire dataset
p=np.sum(df['PlayTennis']=='Yes')
n=np.sum(df['PlayTennis']=='No')
total= n + p
print(p)
print(n)
print(total)

4
3
7


In [221]:
EntropyS= -p/(total)*np.log2(p/(total)) - n/(total)*np.log2(n/(total))
EntropyS

0.9852281360342515

## Choose root node

### Outlook

In [222]:
OutlookS=df[df['Outlook']=='Sunny']
OutlookR=df[df['Outlook']=='Rain']
OutlookO=df[df['Outlook']=='Overcast']

display(OutlookS.loc[:,['Outlook','PlayTennis']])
display(OutlookR.loc[:,['Outlook','PlayTennis']])
display(OutlookO.loc[:,['Outlook','PlayTennis']])

Outlook PlayTennis
0   Sunny         No
1   Sunny         No

Outlook PlayTennis
3    Rain        Yes
4    Rain        Yes
5    Rain         No

Outlook PlayTennis
2  Overcast        Yes
6  Overcast        Yes

In [223]:
#TO AVOID INFINITE LOG WE WILL ADD A VERY SMALL NUMBER TO IT 
eps=0.0000000001

In [224]:
#Entropy for sunny
pS=np.sum(OutlookS['PlayTennis']=='Yes')
nS=np.sum(OutlookS['PlayTennis']=='No')

ESunny=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
ESunny

-1.4426951601859516e-10

In [225]:
#Entropy for rain
pR=np.sum(OutlookR['PlayTennis']=='Yes')
nR=np.sum(OutlookR['PlayTennis']=='No')

ERain=-pR/(pR+nR)*np.log2(pR/(pR+nR) + eps) - nR/(pR+nR)*np.log2(nR/(pR+nR) + eps)
ERain

0.9182958337659504

In [226]:
#Entropy for overcast
pO=np.sum(OutlookO['PlayTennis']=='Yes')
nO=np.sum(OutlookO['PlayTennis']=='No')

EOvercast=-pO/(pO+nO)*np.log2(pO/(pO+nO) + eps) - nO/(pO+nO)*np.log2(nO/(pO+nO) + eps)
EOvercast

-1.4426951601859516e-10

In [227]:
# Calculate average information entropy

IOutlook = (pS+nS)/total*ESunny + (pR+nR)/total*ERain + (pO+nO)/total*EOvercast
IOutlook

0.3935553572458247

In [228]:
# Calculate information gain
IGOutlook=EntropyS-IOutlook
IGOutlook

0.5916727787884268

### Temperature

In [229]:
TempH=df[df['Temperature']=='Hot']
TempM=df[df['Temperature']=='Mild']
TempC=df[df['Temperature']=='Cool']

display(TempH.loc[:,['Temperature','PlayTennis']])
display(TempM.loc[:,['Temperature','PlayTennis']])
display(TempC.loc[:,['Temperature','PlayTennis']])

Temperature PlayTennis
0         Hot         No
1         Hot         No
2         Hot        Yes

Temperature PlayTennis
3        Mild        Yes

Temperature PlayTennis
4        Cool        Yes
5        Cool         No
6        Cool        Yes

In [230]:
#Entropy for hot
pH=np.sum(TempH['PlayTennis']=='Yes')
nH=np.sum(TempH['PlayTennis']=='No')

EHot=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHot

0.9182958337659504

In [231]:
#Entropy for mild
pM=np.sum(TempM['PlayTennis']=='Yes')
nM=np.sum(TempM['PlayTennis']=='No')

EMild=-pM/(pM+nM)*np.log2(pM/(pM+nM) + eps) - nM/(pM+nM)*np.log2(nM/(pM+nM) + eps)
EMild

-1.4426951601859516e-10

In [232]:
#Entropy for cool
pC=np.sum(TempC['PlayTennis']=='Yes')
nC=np.sum(TempC['PlayTennis']=='No')

ECool=-pC/(pC+nC)*np.log2(pC/(pC+nC) + eps) - nC/(pC+nC)*np.log2(nC/(pC+nC) + eps)
ECool

0.9182958337659504

In [233]:
# Calculate average information entropy

ITemp= (pH+nH)/total*EHot + (pM+nM)/total*EMild + (pC+nC)/total*ECool
ITemp

0.7871107146359189

In [234]:
# Calculate information gain
IGTemp=EntropyS-ITemp
IGTemp

0.19811742139833266

### Humidity

In [235]:
HumidH=df[df['Humidity']=='High']
HumidN=df[df['Humidity']=='Normal']

display(HumidH.loc[:,['Humidity','PlayTennis']])
display(HumidN.loc[:,['Humidity','PlayTennis']])

Humidity PlayTennis
0     High         No
1     High         No
2     High        Yes
3     High        Yes

Humidity PlayTennis
4   Normal        Yes
5   Normal         No
6   Normal        Yes

In [236]:
#Entropy for high
pH=np.sum(HumidH['PlayTennis']=='Yes')
nH=np.sum(HumidH['PlayTennis']=='No')

EHigh=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHigh

0.9999999997114609

In [237]:
#Entropy for normal
pN=np.sum(HumidN['PlayTennis']=='Yes')
nN=np.sum(HumidN['PlayTennis']=='No')

ENormal=-pN/(pN+nN)*np.log2(pN/(pN+nN) + eps) - nN/(pN+nN)*np.log2(nN/(pN+nN) + eps)
ENormal

0.9182958337659504

In [238]:
# Calculate average information entropy

IHumid= (pH+nH)/total*EHigh + (pN+nN)/total*ENormal 
IHumid

0.9649839285919564

In [239]:
# Calculate information gain
IGHumid=EntropyS-IHumid
IGHumid

0.02024420744229516

### Wind

In [240]:
WindS=df[df['Wind']=='Strong']
WindW=df[df['Wind']=='Weak']

display(WindS.loc[:,['Wind','PlayTennis']])
display(WindW.loc[:,['Wind','PlayTennis']])

Wind PlayTennis
1  Strong         No
5  Strong         No
6  Strong        Yes

Wind PlayTennis
0  Weak         No
2  Weak        Yes
3  Weak        Yes
4  Weak        Yes

In [241]:
#Entropy for strong
pS=np.sum(WindS['PlayTennis']=='Yes')
nS=np.sum(WindS['PlayTennis']=='No')

EStrong=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
EStrong

0.9182958337659504

In [242]:
#Entropy for weak
pW=np.sum(WindW['PlayTennis']=='Yes')
nW=np.sum(WindW['PlayTennis']=='No')

EWeak=-pW/(pW+nW)*np.log2(pW/(pW+nW) + eps) - nW/(pW+nW)*np.log2(nW/(pW+nW) + eps)
EWeak

0.8112781241705938

In [243]:
# Calculate average information entropy

IWind= (pW+nW)/total*EWeak + (pS+nS)/total*EStrong 
IWind

0.857142856854318

In [244]:
# Calculate information gain
IGWind=EntropyS-IWind
IGWind

0.1280852791799335

As we can see the attribute that gives us the maximum information gain is **Outlook** with a gain of 0.59

Therefore, it will be our root node

In [245]:
display(OutlookS.loc[:,['Outlook','PlayTennis']])
display(OutlookR.loc[:,['Outlook','PlayTennis']])
display(OutlookO.loc[:,['Outlook','PlayTennis']])

Outlook PlayTennis
0   Sunny         No
1   Sunny         No

Outlook PlayTennis
3    Rain        Yes
4    Rain        Yes
5    Rain         No

Outlook PlayTennis
2  Overcast        Yes
6  Overcast        Yes

Since Outlook=Overcast and Outlook= Sunny have only positive and negative examples respectively, we don't need to split the tree anymore at these nodes.Therefore, they become **lead nodes**.  

For Outlook=Rain we must keep splitting using the same procedure.

## Outlook = Rain sub-tree

In [246]:
#Entropy of rain
ERain

0.9182958337659504

In [247]:
OutlookR

Day Outlook Temperature Humidity    Wind PlayTennis
3  D4    Rain        Mild     High    Weak        Yes
4  D5    Rain        Cool   Normal    Weak        Yes
5  D6    Rain        Cool   Normal  Strong         No

In [248]:
total=pR+nR
total

3

### Humidity

In [249]:
HumidH=OutlookR[OutlookR['Humidity']=='High']
HumidN=OutlookR[OutlookR['Humidity']=='Normal']

display(HumidH.loc[:,['Humidity','PlayTennis']])
display(HumidN.loc[:,['Humidity','PlayTennis']])

Humidity PlayTennis
3     High        Yes

Humidity PlayTennis
4   Normal        Yes
5   Normal         No

In [250]:
#Entropy for high
pH=np.sum(HumidH['PlayTennis']=='Yes')
nH=np.sum(HumidH['PlayTennis']=='No')

EHigh=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHigh

-1.4426951601859516e-10

In [251]:
#Entropy for normal
pN=np.sum(HumidN['PlayTennis']=='Yes')
nN=np.sum(HumidN['PlayTennis']=='No')

ENormal=-pN/(pN+nN)*np.log2(pN/(pN+nN) + eps) - nN/(pN+nN)*np.log2(nN/(pN+nN) + eps)
ENormal

0.9999999997114609

In [252]:
# Calculate average information entropy

IHumid= (pH+nH)/total*EHigh + (pN+nN)/total*ENormal 
IHumid

0.6666666664262174

In [253]:
# Calculate information gain
IGHumid=ERain-IHumid
IGHumid

0.25162916733973295

### Wind

In [254]:
WindS=OutlookR[OutlookR['Wind']=='Strong']
WindW=OutlookR[OutlookR['Wind']=='Weak']

display(WindS.loc[:,['Wind','PlayTennis']])
display(WindW.loc[:,['Wind','PlayTennis']])

Wind PlayTennis
5  Strong         No

Wind PlayTennis
3  Weak        Yes
4  Weak        Yes

In [255]:
#Entropy for strong
pS=np.sum(WindS['PlayTennis']=='Yes')
nS=np.sum(WindS['PlayTennis']=='No')

EStrong=-pS/(pS+nS)*np.log2(pS/(pS+nS) + eps) - nS/(pS+nS)*np.log2(nS/(pS+nS) + eps)
EStrong

-1.4426951601859516e-10

In [256]:
#Entropy for weak
pW=np.sum(WindW['PlayTennis']=='Yes')
nW=np.sum(WindW['PlayTennis']=='No')

EWeak=-pW/(pW+nW)*np.log2(pW/(pW+nW) + eps) - nW/(pW+nW)*np.log2(nW/(pW+nW) + eps)
EWeak

-1.4426951601859516e-10

In [257]:
# Calculate average information entropy

IWind= (pW+nW)/total*EWeak + (pS+nS)/total*EStrong 
IWind

-1.4426951601859514e-10

In [258]:
# Calculate information gain
IGWind=ERain-IWind
IGWind

0.9182958339102198

### Temperature

In [259]:
TempH=OutlookR[OutlookR['Temperature']=='Hot']
TempM=OutlookR[OutlookR['Temperature']=='Mild']
TempC=OutlookR[OutlookR['Temperature']=='Cool']

display(TempH.loc[:,['Temperature','PlayTennis']])
display(TempM.loc[:,['Temperature','PlayTennis']])
display(TempC.loc[:,['Temperature','PlayTennis']])

Empty DataFrame
Columns: [Temperature, PlayTennis]
Index: []

Temperature PlayTennis
3        Mild        Yes

Temperature PlayTennis
4        Cool        Yes
5        Cool         No

In [260]:
#Entropy for hot
pH=np.sum(TempH['PlayTennis']=='Yes')
nH=np.sum(TempH['PlayTennis']=='No')

EHot=-pH/(pH+nH)*np.log2(pH/(pH+nH) + eps) - nH/(pH+nH)*np.log2(nH/(pH+nH) + eps)
EHot

C:\Users\Paul\.conda\envs\ml\lib\site-packages\ipykernel_launcher.py:5: RuntimeWarning: invalid value encountered in longlong_scalars
  """


nan

In [261]:
#Entropy for mild
pM=np.sum(TempM['PlayTennis']=='Yes')
nM=np.sum(TempM['PlayTennis']=='No')

EMild=-pM/(pM+nM)*np.log2(pM/(pM+nM) + eps) - nM/(pM+nM)*np.log2(nM/(pM+nM) + eps)
EMild

-1.4426951601859516e-10

In [262]:
#Entropy for cool
pC=np.sum(TempC['PlayTennis']=='Yes')
nC=np.sum(TempC['PlayTennis']=='No')

ECool=-pC/(pC+nC)*np.log2(pC/(pC+nC) + eps) - nC/(pC+nC)*np.log2(nC/(pC+nC) + eps)
ECool

0.9999999997114609

In [263]:
# Calculate average information entropy

ITemp=  (pM+nM)/total*EMild + (pC+nC)/total*ECool
ITemp

0.6666666664262174

In [264]:
# Calculate information gain
IGTemp=ERain-ITemp
IGTemp

0.25162916733973295

The attribute with the maximum gain is **Wind** so the final decision tree will look like this.  

<img src="https://i.gyazo.com/cdf2844673cd7b1238c400de56777585.png">

In [265]:
testdf = pd.DataFrame(columns=["Day","Outlook", "Temperature","Humidity","Wind","PlayTennis"],
                   data=[
                         ['D8','Sunny','Mild','High','Weak','No'],
                         ['D9','Sunny','Cool','Normal','Weak','Yes'],
                         ['D10','Rain','Mild','Normal','Weak','Yes'],
                         ['D11','Sunny','Mild','Normal','Strong','Yes'],
                         ['D12','Overcast','Mild','High','Strong','Yes'],
                         ['D13','Overcast','Hot','Normal','Weak','Yes'],
                         ['D14','Rain','Mild','High','Strong','No']])
testdf

Day   Outlook Temperature Humidity    Wind PlayTennis
0   D8     Sunny        Mild     High    Weak         No
1   D9     Sunny        Cool   Normal    Weak        Yes
2  D10      Rain        Mild   Normal    Weak        Yes
3  D11     Sunny        Mild   Normal  Strong        Yes
4  D12  Overcast        Mild     High  Strong        Yes
5  D13  Overcast         Hot   Normal    Weak        Yes
6  D14      Rain        Mild     High  Strong         No

Correctly classified test examples are D8,D10,D12,D13,D14 so 5 out of 7.  

The tree became way too general.We didn't have enough examples to learn different cases.For example,the fact that there were no examples in the test set with Outlook = Sunny and PlayTennis = Yes ,intuitively makes us think that the training set in small.

# Pruning Strategies

**In this case, and others, there are only a few labelled examples available for training (that is, no additional data
is available for testing or validation). Suggest a concrete pruning strategy, that can be readily embedded in the
algorithm, to avoid over fitting. Explain why you think this strategy should work.**

Since we don't have a lot of data,we wont mention pruning strategies like "Minimum error" since they require evaluation of the tree in a validation data set.  

What we can use is **early-stopping**.  

A method to prevent overfitting is to try and stop the tree-building process early, before it produces leaves with **very small samples**. This heuristic is known as early stopping but is also sometimes known as pre-pruning decision trees.

This method will work on a lot of cases because a leaf node with a small number of samples is very likely to be over-specific which is analogous to over-fitting.